pkg/controller/backup_queue_controller.go

```diff
@@ -0,0 +1,368 @@
+		}
+	})
+	return backupList
+}
+
+// SetupWithManager adds the reconciler to the manager
+func (r *backupQueueReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	// For periodic requeue, only consider Queued backups, order by QueuePosition
+	gp := kube.NewGenericEventPredicate(func(object client.Object) bool {
+		backup := object.(*velerov1api.Backup)
+		return backup.Status.Phase == velerov1api.BackupPhaseQueued
+	})
+
+	s := kube.NewPeriodicalEnqueueSource(r.logger.WithField("controller", constant.ControllerBackupQueue), mgr.GetClient(), &velerov1api.BackupList{}, r.frequency, kube.PeriodicalEnqueueSourceOption{
+		Predicates: []predicate.Predicate{gp},
+		OrderFunc:  queuePositionOrderFunc,
+	})
+
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&velerov1api.Backup{}, builder.WithPredicates(predicate.Funcs{
+			UpdateFunc: func(ue event.UpdateEvent) bool {
+				backup := ue.ObjectNew.(*velerov1api.Backup)
+				return backup.Status.Phase == "" || backup.Status.Phase == velerov1api.BackupPhaseNew
+			},
+			CreateFunc: func(ce event.CreateEvent) bool {
+				backup := ce.Object.(*velerov1api.Backup)
+				return backup.Status.Phase == "" || backup.Status.Phase == velerov1api.BackupPhaseNew
+			},
+			DeleteFunc: func(de event.DeleteEvent) bool {
+				return false
+			},
+			GenericFunc: func(ge event.GenericEvent) bool {
+				return false
+			},
+		})).
+		Watches(
+			&velerov1api.Backup{},
+			handler.EnqueueRequestsFromMapFunc(r.findQueuedBackupsToRequeue),
+			builder.WithPredicates(predicate.Funcs{
+				UpdateFunc: func(ue event.UpdateEvent) bool {
+					oldBackup := ue.ObjectOld.(*velerov1api.Backup)
+					newBackup := ue.ObjectNew.(*velerov1api.Backup)
+					return oldBackup.Status.Phase == velerov1api.BackupPhaseInProgress &&
+						newBackup.Status.Phase != velerov1api.BackupPhaseInProgress ||
+						oldBackup.Status.Phase != velerov1api.BackupPhaseQueued &&
+							newBackup.Status.Phase == velerov1api.BackupPhaseQueued
+				},
+				CreateFunc: func(event.CreateEvent) bool {
+					return false
+				},
+				DeleteFunc: func(de event.DeleteEvent) bool {
+					return false
+				},
+				GenericFunc: func(ge event.GenericEvent) bool {
+					return false
+				},
+			})).
+		WatchesRawSource(s).
+		Named(constant.ControllerBackupQueue).
+		Complete(r)
+}
+
+func (r *backupQueueReconciler) listQueuedBackups(ctx context.Context, ns string) ([]velerov1api.Backup, error) {
+	backupList := &velerov1api.BackupList{}
+	backups := []velerov1api.Backup{}
+	if err := r.Client.List(ctx, backupList, &client.ListOptions{Namespace: ns}); err != nil {
+		r.logger.WithError(err).Error("error listing queued backups")
+		return nil, err
+	}
+	for _, backup := range backupList.Items {
+		if backup.Status.Phase == velerov1api.BackupPhaseQueued {
+			backups = append(backups, backup)
+		}
+	}
+	return backups, nil
+}
+
+func (r *backupQueueReconciler) listEarlierBackups(ctx context.Context, ns string, queuePos int) ([]velerov1api.Backup, int, error) {
```

Comment:
For performance consideration, we can combine `listQueuedBackups` and `listEarlierBackups` into one function and return:
1. earlierBackups in `BackupPhaseQueued` phase and sorted
2. earlierBackups in `BackupPhaseInProgress` and `BackupPhaseReadyToStart`